# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Read adapter file and turn into set of keys.
#' @param adapter_file File with adapter names and sequences in tab-delimited pairs
#' @return Vector of pairs where the 1st entry is the adapter name and the 2nd entry is the adapter sequence
NULL

#' Compute adapter content in reads.
#' 
#' @param infile filepath to fastq sequence
#' @param adapters filepath to adapters
#' @return map object with adapter names as the key and the number of times the adapters appears in the reads as the value
#' @export
calc_adapter_content <- function(infile, adapters) {
    .Call('_qckitfastq_calc_adapter_content', PACKAGE = 'qckitfastq', infile, adapters)
}

#' Read in a BAM or SAM file. For a BAM file package will need to be
#' linked against zlib.
#' 
#' @param fileName path to SAM or BAM file
#' @return 
#' @export
read_bam_or_sam <- function(bamFileName) {
    .Call('_qckitfastq_read_bam_or_sam', PACKAGE = 'qckitfastq', bamFileName)
}

#' calculate summary of quality scores over position
#'
#' Description
#' @param inmat A matrix of score vectors per position
NULL

#' Get Illumina format
#'
#' @param infile  A string giving the path for the fastqfile
#' @param buffer_size An int for the number of lines to keep in memory
#' @examples
#' infile <- system.file("extdata", "10^5_reads_test.fq.gz", package = "qckitfastq")
#' find_format(infile,10000)
#' @return process fastq and evaluate the format of reads
#' @export
find_format <- function(infile, buffer_size, reads_used) {
    .Call('_qckitfastq_find_format', PACKAGE = 'qckitfastq', infile, buffer_size, reads_used)
}

#' Calculate score based on Illumina format
#'
#' @param score  An ascii quality score from the fastq
#' @param score_format The illumina format
#' @examples
#' infile <- system.file("extdata", "10^5_reads_test.fq.gz", package = "qckitfastq")
#' find_format(infile,10000, 100000)
#' @return a string as with the best guess as to the illumina format
#' @export
calc_format_score <- function(score, score_format) {
    .Call('_qckitfastq_calc_format_score', PACKAGE = 'qckitfastq', score, score_format)
}

#' Process fastq and generate sequence and quality score tables
#'
#' @param infile  A string giving the path for the fastqfile
#' @param buffer_size An int for the number of lines to keep in memory
#' @examples
#' infile <- system.file("extdata", "10^5_reads_test.fq.gz", package = "qckitfastq")
#' process_fastq(infile,10000)
#' @return process fastq and generate sequence and quality score tables
#' @export
process_fastq <- function(infile, buffer_size) {
    invisible(.Call('_qckitfastq_process_fastq', PACKAGE = 'qckitfastq', infile, buffer_size))
}

#' Calculate the mean quality score per read of the FASTQ gzipped file
#' 
#' @param infile A string giving the path for the fastqfile
#' @examples
#' infile <- system.file("extdata", "10^5_reads_test.fq.gz", package = "qckitfastq")
#' qual_score_per_read(infile)$q50_per_position[1:10]
#' @return mean quality per read
#' @export
qual_score_per_read <- function(infile) {
    .Call('_qckitfastq_qual_score_per_read', PACKAGE = 'qckitfastq', infile)
}

#' Calculate GC nucleotide sequence content per read of the FASTQ gzipped file
#' @param infile A string giving the path for the fastqfile
#' @examples
#' infile <- system.file("extdata", "10^5_reads_test.fq.gz", package = "qckitfastq")
#' gc_per_read(infile)[1:10]
#' @return GC content perncentage per read
#' @export
gc_per_read <- function(infile) {
    .Call('_qckitfastq_gc_per_read', PACKAGE = 'qckitfastq', infile)
}

#' Calculate sequece counts for each unique sequence and create a table with unique sequences
#' and corresponding counts
#' @param infile A string giving the path for the fastqfile
#' @param min_size An int for thhresholding over representation
#' @param buffer_size An int for the number of lines to keep in memory
#' @return calculate overrepresented sequence count
#' @export
calc_over_rep_seq <- function(infile, min_size = 5L, buffer_size = 1000000L) {
    .Call('_qckitfastq_calc_over_rep_seq', PACKAGE = 'qckitfastq', infile, min_size, buffer_size)
}

